src/MultigridMesh.f90

```diff
@@ -0,0 +1,418 @@
+    !>  into.  May have another use/definition in cases where the interpolation
+    !>  operator has no injections.  This is not needed for fillInterpMatrices
+    !>  in LinearSolverTypes_Multigrid, it is only here for the user's
+    !>  convenience and using/setting this variable is optional.
+    INTEGER(SIK) :: finestGridIndex=-1
+    !> For degree 0 points, childIndices is a size-1 array containing the
+    !>  **child** mesh index of the point that it injects into.
+    !> For degree >0 points, childIndices is a size >1 array containing the
+    !>  **current** mesh indices of the points from which the value on this point
+    !>  is interpolated.
+    !> For MPACT, ordering should be WSENBT on finest mesh and WESNBT on
+    !>  all other meshes, but, in general, ordering is arbitrary
+    INTEGER(SIK),ALLOCATABLE :: childIndices(:)
+
+    CONTAINS
+      !> @copybrief MultigridMesh::clear_MultigridMeshElement
+      !> @copydetails MultigridMesh::clear_MultigridMeshElement
+      PROCEDURE,PASS :: clear => clear_MultigridMeshElement
+  ENDTYPE MultigridMeshElementType
+
+  TYPE :: MultigridMeshType
+    !> Which level the mesh corresponds to (1 = coarsest)
+    INTEGER(SIK) :: iLevel=-1
+    !> Whether or not it is the finest level:
+    LOGICAL(SBK) :: isFinestLevel=.FALSE.
+    !> Whether or not it is the coarsest level:
+    !> NOTE: On the coarsest level, mmData, interpDegrees, and xyzMap need
+    !>  not be allocated!
+    LOGICAL(SBK) :: isCoarsestLevel=.FALSE.
+    !> Number of cells/points on this level locally:
+    INTEGER(SIK) :: nPointsLocal=-1
+    !> Local starting index of mesh points:
+    INTEGER(SIK) :: istt=-1
+    !> Local end index of mesh points:
+    INTEGER(SIK) :: istp=-1
+    !> Global x,y,z location of each element in units of the finest mesh
+    !> This means that adjacent cells in coarser meshes do not have
+    !>   adjacent x/y/z values.  Range is (3,istt:istp)
+    INTEGER(SIK),ALLOCATABLE :: xyzMap(:,:)
+    !> Data for the individual coarse mesh elements. indices are istt:istp
+    TYPE(MultigridMeshElementType),ALLOCATABLE :: mmData(:)
+    !> How many "degrees" each point is from a coarse point.
+    !>  0 = corresponds to a point on the coarse grid
+    !>  1 = between two 0-th degree points
+    !>  2 = between 4 1st degree points, should only exist for 2D/3D problems
+    !>  3 = between 6 2nd degree points, should only exist for 3D problems
+    !>  size of mmData(i)%childIndices or mmData(i)%childWeights is
+    !>    MAX(1,2*interpDegrees(i))
+    !>  indices of interpDegrees should be istt:istp
+    !> Ideally, this would be stored in MultigridMeshElementType, but having
+    !>  it stored as an array is really handy for PETSc's dnnz arguments.
+    INTEGER(SIK),ALLOCATABLE :: interpDegrees(:)
+    !TODO Generalize this logic beyond structured Cartesian grids.
+    !  Actually, I think it should work as is for unstructured grids.
```

Comment:
Pick one?  I'm fine with the TODO as long as it's clear what your expectations are.

Reply:
Good catch -- I clarified the comment to describe the situations in which it should work.